core/routemgmt/activateApi.js

```diff
@@ -0,0 +1,383 @@
+ * https://docs.cloudant.com/document.html#documentCreate
+ *
+ * Parameters (all as fields in the message JSON object)
+ *   host       Required. The database dns host name
+ *   port       Required. The database port number
+ *   protocol   Required. The database protocol (i.e. http, https)
+ *   dbname     Required. The name of the database
+ *   username   Required. The database user name used to access the database
+ *   password   Required. The database user password
+ *   gwUrl      Required. The API Gateway base path (i.e. http://gw.com)
+ *   namespace  Required. The namespace of the API to be activated
+ *   basepath   Required. The basepath of the API to be activated
+ *   docid      Optional. If specified, use this docid to retrieve API doc from DB
+ *
+ * NOTE: The package containing this action will be bound to the following values:
+ *         host, port, protocol, dbname, username, password, gwUrl, gwUser, gwPwd
+ *       As such, the caller to this action should normally avoid explicitly setting
+ *       these values
+ **/
+ var request = require('request');
+ var url = require('url');
+
+function main(message) {
+  var badArgMsg = '';
+  if (badArgMsg = validateArgs(message)) {
+    return whisk.error(badArgMsg);
+  }
+  var dbname = message.dbname;
+
+  var gwInfo = {
+    gwUrl: message.gwUrl,
+  };
+  if (message.gwUser && message.gwPwd) {
+    gwInfo.gwAuth = Buffer.from(message.gwUser+':'+message.gwPwd,'ascii').toString('base64');
+  }
+
+  // Log parameter values
+  console.log('DB host    : '+message.host);
+  console.log('DB port    : '+message.port);
+  console.log('DB protocol: '+message.protocol);
+  console.log('DB username: '+message.username);
+  console.log('DB database: '+message.dbname);
+  console.log('GW URL     : '+message.gwUrl);
+  console.log('GW User    : '+message.gwUser);
+  console.log('GW Pwd     : '+message.gwPwd);
+  console.log('GW Auth    : '+gwInfo.gwAuth);
+  console.log('namespace  : '+message.namespace);
+  console.log('basepath   : '+message.basepath);
+  console.log('docid      : '+message.docid);
+
+  var cloudantOrError = getCloudantAccount(message);
+  if (typeof cloudantOrError !== 'object') {
+    console.error('CloudantAccount returned an unexpected object type: '+(typeof cloudantOrError));
+    return whisk.error('Internal error.  An unexpected object type was obtained.');
+  }
+  var cloudant = cloudantOrError;
+  var cloudantDb = cloudant.use(dbname);
+
+  // Activating a specified API
+  // 1. Pull the API's configuration document from the DB
+  // 2. Create the request document to send to the API Gateway
+  // 3. Send the request to the API Gateway to configure and activate the API
+  // 4. If successful, update the API configuration in the DB to indicate the activation status
+  var retApi;
+  var tenantGuid;
+  var tenantAdded = false;
+  var gwApiActivated = false;
+  var dbUpdated = false;
+  var gwApiDoc;
+  var dbApiDoc;
```

Comment:
I understand why these definitions are here, but this feels like quite brittle to me as you mutate all those variables in Promises down the road. Is there a way to overcome this `undefined` hell?